### PR TITLE
Vault implemented

### PR DESCRIPTION
**Pull request by davidpodhola**

### CODE REVIEW

**Comment by haf** [comment]
Ooh, candy. Will try to look through tomorrow (pretty big ;))

**Comment by davidpodhola** [comment]
Cool, let me please now if anything should be cleared up.

File: src/Fakta.IntegrationTests/GenerateRoot.fs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+open Fakta.Logging
+open Fakta.Vault
+open System.Text
+
+let generateRootInitTest =
+  let otp = 
+    Array.init 16 (fun i -> byte(i*i))
+    |> Convert.ToBase64String
+
+  let listing = GenerateRoot.Init vaultState (("otp", otp), [])
```
Comment (haf) [inline src/Fakta.IntegrationTests/GenerateRoot.fs:19]: Lower-case method names/function names

File: src/Fakta.IntegrationTests/Capabilities.fs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+open NodaTime
+open Fakta
+open Fakta.Logging
+open Fakta.Vault
+
+[<Tests>]
+let tests =
+  testList "Vault capabilities tests" [
+    testCase "sys.capabilities -> capabilities of the token on the given path." <| fun _ ->
+      let map = Map.empty.Add("token", initState.config.token.Value).Add("path", "secret")
```
Comment (haf) [inline src/Fakta.IntegrationTests/Capabilities.fs:17]: (no fix needed) prefer `Map.add` over `(:Map).Add(item)` – module over instance.

File: src/Fakta.IntegrationTests/Audit.fs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+let filePath = @"C:\audit.log"
+let path = "vault-audit"
+
+[<Tests>]
+let tests =
+  testList "Vault audit tests" [
+    testCase "sys.enableAudit -> enable file audit" <| fun _ ->
+      let audit: Audit =
+        {
+            Path = Some(path)
```
Comment (haf) [inline src/Fakta.IntegrationTests/Audit.fs:22]: 2 space indentation

File: src/Fakta.IntegrationTests/Audit.fs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+let filePath = @"C:\audit.log"
+let path = "vault-audit"
+
+[<Tests>]
+let tests =
+  testList "Vault audit tests" [
+    testCase "sys.enableAudit -> enable file audit" <| fun _ ->
+      let audit: Audit =
+        {
```
Comment (haf) [inline src/Fakta.IntegrationTests/Audit.fs:21]: No newline

File: src/Fakta.IntegrationTests/Audit.fs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+open System.Net
+open Chiron
+open Chiron.Operators
+open Fuchu
+open NodaTime
+open Fakta
+open Fakta.Logging
+open Fakta.Vault
+
+let filePath = @"C:\audit.log"
```
Comment (haf) [inline src/Fakta.IntegrationTests/Audit.fs:13]: This won't work on OS X which I'm running

File: src/Fakta.IntegrationTests/HealthVault.fs [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+open Fakta
+open Fakta.Logging
+open Fakta.Vault
+
+
+[<Tests>]
+let tests =
+  testList "Vault health tests" [
+    testCase "sys.vaultHealth -> get health status of vault" <| fun _ ->
+      let listing = Health.GetHealth initState []
```
Comment (haf) [inline src/Fakta.IntegrationTests/HealthVault.fs:18]: camelCase

File: src/Fakta.IntegrationTests/Keys.fs [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+g/Ltg3luWuJ5YM0oRnSfRikIe8QI7SxnjfnIig58lhbT/raHFCv4LleMhqCsKOZT
+ikwDanD67vlmIOvGVbLk+cEbxjTMGb3O8DzFzxkJfoj+g2XdwPTKyxpr/g9b8YAB
+GDpQjznrnJ3Ei6aeYuXjNKOVhEZTpAldWFCVh40ABgsZ4zsv3pICbjScCEBdM0Af
+jsg0HMRLT0pw3aL6aFv2355yiNg9LKvOg31pkBw/N49W3kTeFGRt6HxTqWnlE3e0
+Ml806J728vVeyeOJygHrcThQGAQGLbwyY9bg7Oc5CF/L
+=nyy+
+-----END PGP PUBLIC KEY BLOCK-----"
+
+let base64pgp (pgp : string) =
+    pgp
```
Comment (haf) [inline src/Fakta.IntegrationTests/Keys.fs:32]: 2 space indentation

File: src/Fakta.IntegrationTests/Keys.fs [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    |> Encoding.UTF8.GetBytes
+    |> Convert.ToBase64String
+    
+
+let initRequest: RekeyInitRequest =
+  { SecretShares = 1
```
Comment (haf) [inline src/Fakta.IntegrationTests/Keys.fs:38]: `fieldName` <- standard in this repo

File: src/Fakta.IntegrationTests/Audit.fs (hunk 5 above)
Comment (mavnn) [inline src/Fakta.IntegrationTests/Audit.fs:13]: It's not ideal, but rather unexpectedly it does work. It just writes a file called "c:\audit.log" in the working directory.

File: src/Fakta.IntegrationTests/Keys.fs [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    Backup = None//Some true
+    }
+
+
+
+[<Tests>]
+let tests =
+  let nonce =
+    let listing = Keys.RekeyInit initState (initRequest, [])
+    ensureSuccess listing <| fun (resp) ->
```
Comment (haf) [inline src/Fakta.IntegrationTests/Keys.fs:50]: Unnecessary parenthesis.

File: src/Fakta.IntegrationTests/Mounts.fs [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+open System.Net
+open Chiron
+open Chiron.Operators
+open Fuchu
+open NodaTime
+open Fakta
+open Fakta.Logging
+open Fakta.Vault
+
+let mountConfig: MountConfigInput = { DefaultLeaseTTL = "0"
```
Comment (haf) [inline src/Fakta.IntegrationTests/Mounts.fs:13]: Newline preferred after first `=` for initialisation like this.

File: src/Fakta.IntegrationTests/Mounts.fs [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+                                      MaxLeaseTTL = "0"}
+
+let mountInput: MountInput =
+            { ``Type`` = "consul"
```
Comment (haf) [inline src/Fakta.IntegrationTests/Mounts.fs:17]: Indentation 2 spaces

File: src/Fakta.IntegrationTests/Program.fs [hunk 12]
```diff
@@ -4,13 +4,29 @@ open Fuchu
...
-  //Tests.run ACL.tests
-  //Tests.run Catalog.tests
-  //Tests.run Event.tests
-  //Tests.run Health.tests
-  //Tests.run KV.tests
-  //Tests.run Session.tests
-  //Tests.run Status.tests
+  //Tests.defaultMainThisAssembly argv
+  
+  // I don't know how else to make tests to run in the exact order
```
Comment (haf) [inline src/Fakta.IntegrationTests/Program.fs:9]: Fair enough. What tests depend on what?

You could do

``` fsharp

let tests = [
  Tests.run Agent.tests
  Tests.run ACL.tests
  // ....
  ]
  (tests |>List.fold (fun s t -> s + Tests.run) 0)
```

To keep track of exit codes.

File: src/Fakta.IntegrationTests/TestFactory.fs [hunk 13]
```diff
@@ -24,8 +25,22 @@ let logger =
         printfn "%A" message
     }
 
+let initVault =
+  let reqJson : InitRequest =
+         {secretShares = 1
```
Comment (haf) [inline src/Fakta.IntegrationTests/TestFactory.fs:30]: Indentation and space

File: src/Fakta/Audit.fs [hunk 14]
```diff
@@ -1,2 +1,76 @@
-﻿module Fakta.Vault.Sys.Audit
+﻿module Fakta.Vault.Audit
```
Comment (haf) [inline src/Fakta/Audit.fs:1]: Looks good. Only `functionNamesLeft`

File: src/Fakta/Audit.fs [hunk 15]
```diff
@@ -1,2 +1,76 @@
...
 
+open Fakta
+open Fakta.Logging
+open Fakta.Impl
+open System
+open System.Collections
+open NodaTime
+open HttpFs.Client
+open Chiron
+open Hopac
```
Comment (haf) [inline src/Fakta/Audit.fs:11]: We probably don't need all imports?

File: src/Fakta/Audit.fs [hunk 16]
```diff
@@ -1,2 +1,76 @@
-﻿module Fakta.Vault.Sys.Audit
+﻿module Fakta.Vault.Audit
 
+open Fakta
+open Fakta.Logging
+open Fakta.Impl
+open System
+open System.Collections
+open NodaTime
+open HttpFs.Client
```
Comment (haf) [inline src/Fakta/Audit.fs:9]: Usage is explicit lower towards the bottom.

File: src/Fakta/Audit.fs [hunk 17]
```diff
@@ -1,2 +1,76 @@
...
+    |> withVaultHeader state.config
+
+  let filters =
+    writeFilters state "disableAudit"
+    >> respBodyFilter
+    >> codec createRequest hasNoRespBody
+
+  HttpFs.Client.getResponse |> filters
+
+let HashAudit state : WriteCallNoMeta<string*Map<string,string>, Map<string,string>> =
```
Comment (haf) [inline src/Fakta/Audit.fs:63]: What do I get back from this?

File: src/Fakta/Impl.fs [hunk 18]
```diff
@@ -390,6 +387,15 @@ let inline internal fstOfJsonPrism jsonPrism (item1, item2) : Choice< ^a * 'b, E
     sprintf "Json deserialisation tells us this error: '%s'. Couldn't deserialise input:\n%s" msg item1)
   |> Choice.mapSnd Error.Message
 
+let inline internal fstOfJsonPrismNoMeta jsonPrism (item1) : Choice<'a, Error> =
```
Comment (haf) [inline src/Fakta/Impl.fs:390]: Nice

File: src/Fakta/Keys.fs [hunk 19]
```diff
@@ -1,2 +1,203 @@
...
+open Fakta.Impl
+open System
+open System.Collections
+open NodaTime
+open HttpFs.Client
+open Chiron
+open Hopac
+
+
+let keysPath (funcName: string) =
```
Comment (haf) [inline src/Fakta/Keys.fs:14]: internal or private modifier needed?

File: src/Fakta/Init.fs [hunk 20]
```diff
@@ -1,2 +1,48 @@
...
+open Fakta.Impl
+open System
+open System.Collections
+open NodaTime
+open HttpFs.Client
+open Chiron
+open Hopac
+
+
+let initPath (funcName: string) =
```
Comment (haf) [inline src/Fakta/Init.fs:14]: internal or private modifier needed?

File: src/Fakta/HealthVault.fs [hunk 21]
```diff
@@ -1,2 +1,32 @@
...
+open Fakta.Impl
+open System
+open System.Collections
+open NodaTime
+open HttpFs.Client
+open Chiron
+open Hopac
+
+
+let healthPath (funcName: string) =
```
Comment (haf) [inline src/Fakta/HealthVault.fs:14]: internal or private modifier needed?

File: src/Fakta/GenerateRoot.fs [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+open Fakta.Impl
+open System
+open System.Collections
+open NodaTime
+open HttpFs.Client
+open Chiron
+open Hopac
+
+
+let genRootPath (funcName: string) =
```
Comment (haf) [inline src/Fakta/GenerateRoot.fs:14]: internal or private modifier needed?

File: src/Fakta/Capabilities.fs [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+open Fakta.Impl
+open System
+open System.Collections
+open NodaTime
+open HttpFs.Client
+open Chiron
+open Hopac
+
+
+let capabilitiesPath (funcName: string) =
```
Comment (haf) [inline src/Fakta/Capabilities.fs:14]: internal or private modifier needed?

File: src/Fakta/Auth.fs [hunk 24]
```diff
@@ -1,2 +1,62 @@
...
+open Fakta.Impl
+open System
+open System.Collections
+open NodaTime
+open HttpFs.Client
+open Chiron
+open Hopac
+
+
+let authPath (funcName: string) =
```
Comment (haf) [inline src/Fakta/Auth.fs:14]: internal or private modifier needed?

File: src/Fakta/Audit.fs [hunk 25]
```diff
@@ -1,2 +1,76 @@
...
+open Fakta.Impl
+open System
+open System.Collections
+open NodaTime
+open HttpFs.Client
+open Chiron
+open Hopac
+
+
+let auditPath (funcName: string) =
```
Comment (haf) [inline src/Fakta/Audit.fs:14]: internal or private modifier needed?

File: src/Fakta/Leader.fs [hunk 26]
```diff
@@ -0,0 +1,33 @@
...
+open Fakta.Impl
+open System
+open System.Collections
+open NodaTime
+open HttpFs.Client
+open Chiron
+open Hopac
+
+
+let leaderPath (funcName: string) =
```
Comment (haf) [inline src/Fakta/Leader.fs:14]: internal or private modifier needed?

File: src/Fakta/Leases.fs [hunk 27]
```diff
@@ -1,2 +1,63 @@
...
+open Fakta.Impl
+open System
+open System.Collections
+open NodaTime
+open HttpFs.Client
+open Chiron
+open Hopac
+
+
+let leasesPath (funcName: string) =
```
Comment (haf) [inline src/Fakta/Leases.fs:14]: internal or private modifier needed?

File: src/Fakta/Leases.fs [hunk 28]
```diff
@@ -1,2 +1,63 @@
...
+  [| "Fakta"; "Vault"; "Sys"; funcName |]
+
+let queryFilters state =
```
Comment (haf) [inline src/Fakta/Leases.fs:17]: internal or private modifier needed?

File: src/Fakta/Leases.fs [hunk 29]
```diff
@@ -1,2 +1,63 @@
...
+  leasesPath >> queryFiltersNoMeta state
+
+let writeFilters state =
```
Comment (haf) [inline src/Fakta/Leases.fs:20]: internal or private modifier needed?

File: src/Fakta/Policy.fs [hunk 30]
```diff
@@ -1,2 +1,73 @@
...
+open Fakta.Impl
+open System
+open System.Collections
+open NodaTime
+open HttpFs.Client
+open Chiron
+open Hopac
+
+
+let policyPath (funcName: string) =
```
Comment (haf) [inline src/Fakta/Policy.fs:14]: internal or private modifier needed?

File: src/Fakta/Types.fs [hunk 31]
```diff
@@ -1196,14 +1196,26 @@ type FaktaConfig =
...
+  static member ConsulEmpty =
     { serverBaseUri  = Uri "http://127.0.0.1:8500"
       datacenter     = None
       credentials    = None
       waitTime       = DefaultLockWaitTime
-      token          = None }
+      token          = None 
+      keys           = None}
+
+  static member VaultEmpty (token: Token) (keys: Token list) =
```
Comment (haf) [inline src/Fakta/Types.fs:1210]: Naming standard, `vaultEmpty`

File: src/Fakta/Types.fs [hunk 32]
```diff
@@ -1196,14 +1196,26 @@ type FaktaConfig =
     waitTime      : Duration
     /// Token is used to provide a per-request ACL token
     /// which overrides the agent's default token.
-    token         : Token option }
+    token         : Token option
+    keys          : Token list option }
 
-  static member empty =
+  static member ConsulEmpty =
```
Comment (haf) [inline src/Fakta/Types.fs:1202]: naming standard `consulEmpty`. Breaking change (ok)

File: src/Fakta/Types.fs [hunk 33]
```diff
@@ -1214,15 +1226,585 @@ let random =
...
 type FaktaState =
   { config : FaktaConfig
     logger : Logger
     clock  : IClock
     random : Random }
 
-  static member empty =
-    { config = FaktaConfig.empty
-      logger = NoopLogger
+  static member empty (api: APIType) (token: Token) (keys: Token list) (logger: Logger) =
```
Comment (haf) [inline src/Fakta/Types.fs:1239]: Empty takes **no** parameters. Create new `create` function for this use-case.

File: src/Fakta/Types.fs [hunk 34]
```diff
@@ -1214,15 +1226,585 @@ let random =
...
+    <*> Json.read "progress"
+
+  static member ToJson (se : SealStatusResponse) =
+    Json.write "sealed" se.``sealed``
+    *> Json.write "t" se.t
+    *> Json.write "n" se.n
+    *> Json.write "progress" se.progress
+
+type GenerateRootStatusResponse =
+  { Nonce   : string
```
Comment (haf) [inline src/Fakta/Types.fs:1313]: Naming standard.

File: src/Fakta/Types.fs [hunk 35]
```diff
@@ -1214,15 +1226,585 @@ let random =
...
+    Json.write "nonce" se.Nonce
+    *> Json.write "started" se.Started
+    *> Json.write "progress" se.Progress
+    *> Json.write "required" se.Required
+    *> Json.write "complete" se.Complete
+    *> Json.write "encoded_root_token" se.EncodedRootToken
+    *> Json.write "pgp_fingerprint" se.PGPFingerprint
+
+type MountConfigInput =
+  { DefaultLeaseTTL : string
```
Comment (haf) [inline src/Fakta/Types.fs:1349]: Naming standard.

File: src/Fakta/Types.fs [hunk 36]
```diff
@@ -1214,15 +1226,585 @@ let random =
...
+    <!> Json.read "default_lease_ttl"
+    <*> Json.read "max_lease_ttl"
+    
+
+  static member ToJson (se : MountConfigInput) =
+    Json.write "default_lease_ttl" se.DefaultLeaseTTL
+    *> Json.write "max_lease_ttl" se.MaxLeaseTTL
+
+type MountConfigOutput =
+  { DefaultLeaseTTL : int
```
Comment (haf) [inline src/Fakta/Types.fs:1365]: Naming standard.

File: src/Fakta/Types.fs [hunk 37]
```diff
@@ -1214,15 +1226,585 @@ let random =
...
+    <!> Json.read "default_lease_ttl"
+    <*> Json.read "max_lease_ttl"
+    
+
+  static member ToJson (se : MountConfigOutput) =
+    Json.write "default_lease_ttl" se.DefaultLeaseTTL
+    *> Json.write "max_lease_ttl" se.MaxLeaseTTL
+
+type MountOutput =
+  { ``Type``    : string
```
Comment (haf) [inline src/Fakta/Types.fs:1381]: Naming standard.

File: src/Fakta/Types.fs [hunk 38]
```diff
@@ -1214,15 +1226,585 @@ let random =
...
+    <*> Json.tryRead "config"
+    
+
+  static member ToJson (se : MountOutput) =
+    Json.write "type" se.Type
+    *> Json.write "description" se.Description
+    *> Json.maybeWrite "config" se.Config
+
+type MountInput =
+  { ``Type``    : string
```
Comment (haf) [inline src/Fakta/Types.fs:1401]: Naming standard.

File: src/Fakta/Types.fs [hunk 39]
```diff
@@ -1214,15 +1226,585 @@ let random =
...
+    
+
+  static member ToJson (se : MountInput) =
+    Json.write "type" se.Type
+    *> Json.write "description" se.Description
+    *> Json.maybeWrite "config" se.Config
+
+
+type LeaderResponse =
+  { HAEnabled     : bool
```
Comment (haf) [inline src/Fakta/Types.fs:1422]: Naming standard.

File: src/Fakta/Types.fs [hunk 40]
```diff
@@ -1214,15 +1226,585 @@ let random =
...
+    <*> Json.read "leader_address"
+    
+
+  static member ToJson (se : LeaderResponse) =
+    Json.write "ha_enabled" se.HAEnabled
+    *> Json.write "is_self" se.IsSelf
+    *> Json.write "leader_address" se.LeaderAddress
+
+type SecretWrapInfo = 
+  { Token         : Token
```
Comment (haf) [inline src/Fakta/Types.fs:1442]: Naming standard.

File: src/Fakta/Types.fs [hunk 41]
```diff
@@ -1214,15 +1226,585 @@ let random =
...
+    <*> Json.read "creation_time"
+    
+
+  static member ToJson (se : SecretWrapInfo) =
+    Json.write "token" se.Token
+    *> Json.write "ttl" se.TTL
+    *> Json.write "creation_time" se.CreationTime
+
+type SecretAuth = 
+  { ClientToken   : Token
```
Comment (haf) [inline src/Fakta/Types.fs:1464]: Naming standard.

File: src/Fakta/Types.fs [hunk 42]
```diff
@@ -1214,15 +1226,585 @@ let random =
...
+  static member ToJson (se : SecretAuth) =
+    Json.write "client_token" se.ClientToken
+    *> Json.write "accessor" se.Accessor
+    *> Json.write "policies" se.Policies
+    *> Json.write "metadata" se.Metadata
+    *> Json.write "lease_duration" se.LeaseDuration
+    *> Json.write "renewable" se.Renewable
+
+type SecretDataList =
+  { LeaseId       : string
```
Comment (haf) [inline src/Fakta/Types.fs:1497]: Naming standard.

File: src/Fakta/Types.fs [hunk 43]
```diff
@@ -1214,15 +1226,585 @@ let random =
...
+    Json.write "lease_id" se.LeaseId
+    *> Json.write "lease_duration" se.LeaseDuration
+    *> Json.write "renewable" se.Renewable
+    *> Json.write "data" se.Data
+    *> Json.maybeWrite "warnings" se.Warnings
+    *> Json.maybeWrite "auth" se.Auth
+    *> Json.maybeWrite "wrap_info" se.WrapInfo
+
+type SecretDataString =
+  { LeaseId       : string
```
Comment (haf) [inline src/Fakta/Types.fs:1533]: Naming standard.

File: src/Fakta/Types.fs [hunk 44]
```diff
@@ -1214,15 +1226,585 @@ let random =
...
+    *> Json.write "lease_duration" se.LeaseDuration
+    *> Json.write "renewable" se.Renewable
+    *> Json.write "data" se.Data
+    *> Json.maybeWrite "warnings" se.Warnings
+    *> Json.maybeWrite "auth" se.Auth
+    *> Json.maybeWrite "wrap_info" se.WrapInfo
+
+
+type KeyStatus =
+  { Term        : int
```
Comment (haf) [inline src/Fakta/Types.fs:1570]: Naming standard.

File: src/Fakta/Types.fs [hunk 45]
```diff
@@ -1214,15 +1226,585 @@ let random =
...
+    <*> Json.read "install_time"
+    
+
+  static member ToJson (se : KeyStatus) =
+    Json.write "term" se.Term
+    *> Json.write "install_time" se.InstallTime
+
+
+type RekeyInitRequest =
+  { SecretShares    : int
```
Comment (haf) [inline src/Fakta/Types.fs:1587]: Naming standard.

File: src/Fakta/Types.fs [hunk 46]
```diff
@@ -1214,15 +1226,585 @@ let random =
...
+
+  static member ToJson (se : RekeyInitRequest) =
+    Json.write "secret_shares" se.SecretShares
+    *> Json.write "secret_threshold" se.SecretTreshold
+    *> Json.maybeWrite "pgp_keys" se.PGPKeys
+    *> Json.maybeWrite "backup" se.Backup
+
+
+type RekeyStatusResponse =
+  { Nonce           : string
```
Comment (haf) [inline src/Fakta/Types.fs:1612]: Naming standard.

File: src/Fakta/Types.fs [hunk 47]
```diff
@@ -1214,15 +1226,585 @@ let random =
...
+    *> Json.write "started" se.Started
+    *> Json.write "t" se.T
+    *> Json.write "n" se.N
+    *> Json.write "progress" se.Progress
+    *> Json.write "required" se.Required
+    *> Json.maybeWrite "pgp_fingerprints" se.PGPFIngerPrints
+    *> Json.write "backup" se.Backup
+
+type RekeyUpdateResponse =
+  { Nonce           : string
```
Comment (haf) [inline src/Fakta/Types.fs:1652]: Naming standard.

File: src/Fakta/Types.fs [hunk 48]
```diff
@@ -1214,15 +1226,585 @@ let random =
...
+
+  static member ToJson (se : RekeyUpdateResponse) =
+    Json.write "nonce" se.Nonce
+    *> Json.write "complete" se.Complete
+    *> Json.write "keys" se.Keys
+    *> Json.maybeWrite "pgp_fingerprints" se.PGPFIngerPrints
+    *> Json.write "backup" se.Backup
+
+type RekeyRetrieveResponse =
+  { Nonce : string
```
Comment (haf) [inline src/Fakta/Types.fs:1680]: Naming standard.

File: src/Fakta/Types.fs [hunk 49]
```diff
@@ -1214,15 +1226,585 @@ let random =
...
+    <*> Json.read "keys"
+    
+
+  static member ToJson (se : RekeyRetrieveResponse) =
+    Json.write "nonce" se.Nonce
+    *> Json.write "keys" se.Keys
+
+
+type Audit =
+  { Path        : string option
```
Comment (haf) [inline src/Fakta/Types.fs:1697]: Naming standard.

File: src/Fakta/Types.fs [hunk 50]
```diff
@@ -1214,15 +1226,585 @@ let random =
...
+    
+
+  static member ToJson (se : Audit) =
+    Json.maybeWrite "path" se.Path
+    *> Json.write "type" se.Type
+    *> Json.write "description" se.Description
+    *> Json.maybeWrite "options" se.Options
+
+type HealthResponse =
+  { Initialized        : bool
```
Comment (haf) [inline src/Fakta/Types.fs:1721]: Naming standard.

Comment (haf) [inline src/Fakta/Types.fs:1721]: `:` indentation

File: src/Fakta/Types.fs [hunk 51]
```diff
@@ -1214,15 +1226,585 @@ let random =
...
+    
+
+  static member ToJson (se : HealthResponse) =
+    Json.write "initialized" se.Initialized
+    *> Json.write "sealed" se.Sealed
+    *> Json.write "standby" se.Standby
+    *> Json.maybeWrite "server_time_utc" se.ServerTimeUtc
+
+type AuthConfig =
+  { DefaultLeaseTTL        : string
```
Comment (haf) [inline src/Fakta/Types.fs:1745]: Naming standard.

File: src/Fakta/Types.fs [hunk 52]
```diff
@@ -1214,15 +1226,585 @@ let random =
...
+        MaxLeaseTTL = o})
+    <!> Json.read "default_lease_ttl"
+    <*> Json.read "max_lease_ttl"
+
+  static member ToJson (se : AuthConfig) =
+    Json.write "default_lease_ttl" se.DefaultLeaseTTL
+    *> Json.write "max_lease_ttl" se.MaxLeaseTTL
+
+type AuthMount =
+  { Type        : string
```
Comment (haf) [inline src/Fakta/Types.fs:1760]: Naming standard.

File: src/Fakta/Types.fs [hunk 53]
```diff
@@ -1214,15 +1226,585 @@ let random =
...
+    
+    
+
+  static member ToJson (se : AuthMount) =
+    Json.write "type" se.Type
+    *> Json.write "description" se.Description
+    *> Json.maybeWrite "config" se.Config
+
+type AuthMethod =
+  | AppID
```
Comment (haf) [inline src/Fakta/Types.fs:1781]: Looking forward to putting this DU to use IRL

**Comment by haf** [comment]
- ✅ All tests pass on OS X
- ✅ Reviewing it done. 
- ☑️ Really minor code formatting fixes and naming fixes required.

Well done!

**Pull request merged**